Best Practices for reliable automation of UIDT-1 (#25)
askui/docs-mintlify: introduction/03-core-concepts/05-best-practice-for-reliable-automation.mdx

| Expected (Truth) | OCR Output (Prediction) |
| --- | --- |
| `Hallo` | `HaII0` |

### Solutions

### 1. Re-Teach the OCR Model Using AskUI’s OCR Re-Teaching App

You can directly correct OCR predictions and improve model accuracy by training your workspace-specific model.

### Steps:

1. Start the AskUI shell:
    
    ```bash
    bash
    
    askui-shell
    ```
    
2. Launch the OCR Teaching App:
    
    ```bash
    bash
    
    AskUI-StartOCRTeaching
    ```
    
3. Upload a screenshot containing the misclassified word (e.g., “Hallo”).
4. Switch to **Character-Level Mode** for precise corrections.
5. Select the wrongly detected word (`HaII0`) and replace it with the correct label: `Hallo`.
6. Click **"Copy Model"** to copy the newly trained model ID.
7. In your automation code (e.g., `askui-helper.ts`), update the `exec()` call to use the new model:

```
ts
CopyEdit
await aui.click().text("Hallo").exec([{
    "task": "e2e_ocr",
    "architecture": "easy_ocr",
    "version": "1",
    "interface": "online_learning",
    "useCase": "<your-workspace-id>",
    "tags": ["char_level"]
}])

```

## Text Detection Issues

### **Icon Text Merging**

> **Problem:** Sometimes, Text Detector/ annotation tool, **merges an icon and texts into one**, even though they look merged on screen.

Reviewer: ```suggestion
**Problem:** Sometimes, text detector/annotation tool, **merges an icon and texts into one**, even though they look merged on screen.

```

> **Problem:** Sometimes, Text Detector/ annotation tool, **merges an icon and texts into one**, even though they look merged on screen.

Reviewer: ```suggestion
**Problem:** Sometimes, text detector/annotation tool, **merges an icon and texts into one**, even though they look merged on screen.

```

**Example:** Say you want to click **just the name** “Alice Johnson” field or **just the position** field in a interface - but OCR detects them as one long string:

**✅ Expected Behavior**

🖼️ *Icon and Text are detected separately:*

![image.png](attachment:d0cea22a-f716-4e46-b3ca-07866a51513a:image.png)

🧑 ✅ Name ✅  🤖 ✅ Role ✅

👍 Works with `click().text("Name")` or `click().text("Name")`

**❌ Actual Issue**

🖼️ I*con and text are detected together:*

![image.png](attachment:4c1589c9-31af-46bc-aa81-8530abefadf5:image.png)

`🧑 Name` ❌  🤖 ✅ Role ✅

👎 Can't find `click().text("Name")`.

**Solutions:** 

> ### Option 1: Re-Teach the OCR Model to ignore the icon.

Reviewer: ```suggestion
### Option 1: Reteach the OCR Model to ignore the icon.

```

> 1. Start Re-Teachting App

Reviewer: ```suggestion
1. Start reteaching app

```

> 2. Teach the Text Recognition to ignore the icon

Reviewer: ```suggestion
2. Teach the text recognition to ignore the icon

```

> 3. Use the Custom Model.

Reviewer: ```suggestion
3. Use the custom model.

```

```yaml
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "<your-workspace-id>",
        "tags": ["word_level"]
    }])

```

### Option 2: Use Default Word-Level Detection (Best Practice)

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "00000000_0000_0000_0000_000000000000",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.
> 

### Option 3: Use Custom Model Word-Level Detection

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "<your-workspace-id>",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.y.
> 

### Merged Texts

> **Problem:** Sometimes, Text Detector/ annotation tool, **merges two separate texts into one**, even though they look clearly split on screen.

Reviewer: ```suggestion
**Problem:** Sometimes, text detector/annotation tool, **merges two separate texts into one**, even though they look clearly split on screen.

```

![image.png](attachment:38923942-c474-4d0a-89ed-350cc168034b:image.png)

`Alice Johnson Software Engineer`❌

👎 Can't find either one on its own.

**Solutions:** 

### Option 1: Use Default Word-Level Detection (Best Practice)

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "00000000_0000_0000_0000_000000000000",
        "tags": ["word_level"]
    }])
```

> Breaks up merged strings into individual words.
> 

### Option 2: Use Custom Model Word-Level Detection

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "<your-workspace-id>",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.
> 

### Option 3: Use Relative Positioning (Fallback)

```tsx

await aui.moveMouseRelativeTo(0, left).containsText("Name").exec()

```

> Useful when text detection still merges blocks. Navigates close enough to act anyway.
> 

### **Text Seperation**

> **Problem:** Sometimes, Text Detector/ annotation tool, septerates a text into **two texts**, even though they look clearly merged on screen.

Reviewer: ```suggestion
**Problem:** Sometimes, text detector/annotation tool, septerates a text into **two texts**, even though they look clearly merged on screen.

```

![image.png](attachment:263d543a-fa41-42ee-8c33-73db69d84525:image.png)

`Alice Johnson` ✅ 

👍 Works with `text("Alice Johnson")`

**❌ Actual Issue**

🖼️ Words *are detected as seperated texts:*

![image.png](attachment:1e50bb77-6b65-4f95-a16f-22baf51cb0ec:image.png)

`Alice`❌ `Johnson`❌

👎 Can't find either `text("Alice Johnson")` on its own.

**Solutions:** 

### Option 1: Use Default Word-Level Detection (Best Practice)

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "00000000_0000_0000_0000_000000000000",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.
> 

### Option 2: Use Custom Model Word-Level Detection

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "<your-workspace-id>",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.
> 

### **Vertical Text Merging**

> **Problem:** Sometimes, Text Detector/ annotation tool, merges two lines to one text, even though they look clearly as two lines on screen.

Reviewer: ```suggestion
**Problem:** Sometimes, text detector/annotation tool, merges two lines to one text, even though they look clearly as two lines on screen.

```

![image.png](attachment:43f0d240-8244-46db-a197-14970dae1a42:image.png)

`Alice Johnson` ✅ 

👍 Works with `text("Alice Johnson")`

**❌ Actual Issue**

🖼️ Texts *are detected as one text:*

![image.png](attachment:9df90008-3f78-40cb-b9fa-ecb1288e5c93:image.png)

`<no words recognized>`❌

👎 Can't find either `text("Alice Johnson")` on its own.

**Solutions:** 

### Option 1: Use Default Word-Level Detection (Best Practice)

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "00000000_0000_0000_0000_000000000000",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.
> 

### Option 2: Use Custom Model Word-Level Detection

```tsx
await aui.click().text("Alice Johnson").exec([{
        "task": "e2e_ocr",
        "architecture": "easy_ocr",
        "version": "1",
        "interface": "online_learning",
        "useCase": "<your-workspace-id>",
        "tags": ["word_level"]
    }])

```

> Breaks up merged strings into individual words.
> 

### **Single Character not Detected**

> **Problem:** Sometimes, Text Detector/ annotation tool, does not detect single charactors, even though they look clearly on screen.

Reviewer: ```suggestion
**Problem:** Sometimes, text detector/annotation tool, does not detect single charactors, even though they look clearly on screen.

```
